python/mlx/nn/losses.py

```diff
@@ -32,20 +32,59 @@ def cross_entropy(
 
     Args:
         logits (array): The unnormalized predicted logits.
-        targets (array): The target values, as class indices.
-        weights (array, optional): Weights for each target. Default: ``None``.
+        targets (array): The ground truth values. These can be class indices or probabilities for
+            each class, depending on its shape. If ``targets`` are class indices, then the shape of
+            ``targets`` should match the shape of ``logits``, excluding the ``axis`` dimension.
+            If ``targets`` are probabilities (or one-hot encoded), then the shape of ``targets``
+            should be identical to the shape of ``logits``.
+        weights (array, optional): Optional weights for each target. Default: ``None``.
         axis (int, optional): The axis over which to compute softmax. Default: ``-1``.
         label_smoothing (float, optional): Label smoothing factor. Default: ``0``.
         reduction (str, optional): Specifies the reduction to apply to the output:
             ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
 
     Returns:
         array: The computed cross entropy loss.
+
+    Examples:
+        >>> import mlx.core as mx
+        >>> from mlx.nn.losses import cross_entropy
+
+        >>> # Class indices as targets
+        >>> logits = mx.array([[2.0, -1.0], [-1.0, 2.0]])
+        >>> targets = mx.array([0, 1])
+        >>> cross_entropy(logits, targets)
+        array([0.0485873, 0.0485873], dtype=float32)
+
+        >>> # Probabilities (or one-hot vectors) as targets
+        >>> logits = mx.array([[2.0, -1.0], [-1.0, 2.0]])
+        >>> targets = mx.array([[0.9, 0.1], [0.1, 0.9]])
+        >>> cross_entropy(logits, targets)
+        array([0.348587, 0.348587], dtype=float32)
     """
     if label_smoothing < 0 or label_smoothing >= 1:
         raise ValueError(f"Label smoothing must in [0, 1), got {label_smoothing}.")
 
-    score = mx.take_along_axis(logits, targets[..., None], axis).squeeze(-1)
+    # Whether targets are class indices or probabilities
+    targets_as_probs = targets.ndim == logits.ndim
+
+    def _drop_dim(shape, axis):
+        shape.pop(axis)
+        return shape
+
+    # Check shapes in two cases: targets as class indices and targets as probabilities
+    if (targets_as_probs and targets.shape != logits.shape) or (
+        not targets_as_probs and targets.shape != _drop_dim(logits.shape, axis)
+    ):
+        raise ValueError(
+            f"Targets shape {targets.shape} does not match logits shape {logits.shape}."
+        )
+
+    if targets_as_probs:
+        score = mx.sum(logits * targets, axis=axis)  # mx.inner doesn't support axis
```

Comment:
nit: `mx.inner` is unlikely to support axis since numpy is the same, so the comment is not necessary

Reply:
You could (and probably should) use `mx.tensordot` though.

Reply:
Thanks for your advice, but I cannot figure out how to use `mx.tensordot` to do `mx.sum(a * b, axis=axis)`. For example, if both `a` and `b` has shape `[2, 3, 4]`, with `mx.sum(a * b, axis=1)` I can get the desired output with shape `[2, 4]` (`3` is reduced). But with `mx.tensordot(a, b, axes=[[1], [1]])`, it produces shape `[2, 4, 2, 4]`. I would like to use something like `np.einsum` to do `mx.einsum("ijk,ijk->ik", a, b)` but MLX doesn't have a Einsum.

Reply:
Indeed, sorry `tensordot` is not the right call, sorry for the bad suggestion. Let's keep the sum reduce as you have it actually, that is simple and likely fine perf wise.